docs/configuration.md
# Configuration

## Environment Variables

| Variable              | Description |
|-----------------------|-------------|
| CONFIG                | Set the config file path |
| WIREGUARD_PRIVATE_KEY | Set the wireguard private key |
| STORAGE               | Set the directory where device config will be persisted |
| ADMIN_USERNAME        | Set the username (subject) for the admin account |
| ADMIN_PASSWORD        | Set the admin account's password. The admin account will be a basic-auth user. Leave blank if your admin username authenticates via a configured authentication backend. |
| UPSTREAM_DNS          | Set the upstream DNS server to proxy client DNS requests to. If empty, resolv.conf will be respected. |
| LOG_LEVEL             | Set the server's log level (debug, **info**, error, critical) |
| DISABLE_METADATA      | If true, the server will not record device level metadata such as the last handshake time, tx/rx data size |

## CLI Flags

All environment variables can be configured via a
CLI flag as well.

For example you can configure `STORAGE` by passing `--storage="<value>"`.

## Config File (config.yaml)

Here's an annotated config file example:

```yaml
# The application's log level.
# Can be debug, info, error
# Optional, defaults to info
loglevel: info
# Disable device metadata storage.
# Device metadata includes the last handshake time,
# total sent/received bytes count, their endpoint IP.
# This metadata is captured from wireguard itself.
# Disabling this flag will not stop wireguard from capturing
# this data.
# See stored data here: https://github.com/Place1/wg-access-server/blob/master/internal/storage/contracts.go#L14
# Optional, defaults to false.
disableMetadata: false
# The port that the web ui server (http) will listen on.
# Optional, defaults to 8000
port: 8000
# Directory that VPN devices (WireGuard peers)
# What type of storage do you want? inmemory (default), file:///some/directory, or postgres, mysql, sqlite3
storage: "memory://"
wireguard:
  # The network interface name for wireguard
  # Optional, defaults to wg0
  interfaceName: wg0
  # The WireGuard PrivateKey
  # You can generate this value using "$ wg genkey"
  # If this value is empty then the server will use an in-memory
  # generated key
  privateKey: ""
  # ExternalAddress is the address (without port) that clients use to connect to the wireguard interface
  # By default, this will be empty and the web ui
  # will use the current page's origin i.e. window.location.origin
  # Optional
  externalHost: ""
  # The WireGuard ListenPort
  # Optional, defaults to 51820
  port: 51820
vpn:
  # CIDR configures a network address space
  # that client (WireGuard peers) will be allocated
  # an IP address from.
  # Optional
  cidr: "10.44.0.0/24"
  # GatewayInterface will be used in iptable forwarding
  # rules that send VPN traffic from clients to this interface
  # Most use-cases will want this interface to have access
  # to the outside internet
  # If not configured then the server will select the default
  # network interface e.g. eth0
  # Optional
  gatewayInterface: ""
  # The "AllowedIPs" for VPN clients.
  # This value will be included in client config
  # files and in server-side iptable rules
  # to enforce network access.
  # Optional
  allowedIPs:
    - "0.0.0.0/0"
dns:
  # Enable a DNS proxy for VPN clients.
  # Optional, Defaults to true
  enabled: true
  # upstream DNS servers.
  # that the server-side DNS proxy will forward requests to.
  # By default /etc/resolv.conf will be used to find upstream
  # DNS servers.
  # Optional
  upstream:
    - "1.1.1.1"
# Auth configures optional authentication backends
# to controll access to the web ui.
# Devices will be managed on a per-user basis if any
# auth backends are configured.
# If no authentication backends are configured then
# the server will not require any authentication.
# It's recommended to make use of basic authentication
# or use an upstream HTTP proxy that enforces authentication
# Optional
auth:
  # HTTP Basic Authentication
  basic:
    # Users is a list of htpasswd encoded username:password pairs
    # supports BCrypt, Sha, Ssha, Md5
    # You can create a user using "htpasswd -nB <username>"
    users: []
  oidc:
    name: "" # anything you want
    issuer: "" # Should point to the oidc url without .well-known
    clientID: ""
    clientSecret: ""
    scopes: null  # list of scopes, defaults to ["openid"]
    redirectURL: "" # full url you want the oidc to redirect to, example: https://vpn-admin.example.com/finish-signin
    # See https://github.com/Knetic/govaluate/blob/9aa49832a739dcd78a5542ff189fb82c3e423116/MANUAL.md for how to write rules
    userClaimsRules:
      admin: "'WireguardAdmins' in group_membership"
    # Optionally restrict login to users with an allowed email domain
    # if empty or omitted, any email domain will be allowed.
    emailDomains:
      - example.com
  gitlab:
    name: ""
    baseURL: ""
    clientID: ""
    clientSecret: ""
    redirectURL: ""
    # Optionally restrict login to users with an allowed email domain
    # if empty or omitted, any email domain will be allowed.
    emailDomains:
      - example.com
```


docs/2-configuration.md
# Configuration

You can configure wg-access-server using environment variables, cli flags or a config file
taking precedence over one another in that order.

The default configuration should work out of the box if you're just looking to try it out.

The only required configuration is an admin password and a wireguard private key. The admin
password can be anything you like. You can generate a wireguard private key by
[following the official docs](https://www.wireguard.com/quickstart/#key-generation).

TLDR:

```bash
wg genkey
```

The config file format is `yaml` and an example is provided [below](#the-config-file-configyaml).

Here's what you can configure:

| Environment Variable       | CLI Flag                  | Config File Path       | Required | Default (docker)                        | Description                                                                                                                                                                                 |
| -------------------------- | ------------------------- | ---------------------- | -------- | --------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `WG_CONFIG`                | `--config`                | `loglevel`             |          | `info`                                  | Global log level                                                                                                                                                                            |
| `WG_ADMIN_USERNAME`        | `--admin-username`        | `adminUsername`        |          | `admin`                                 | The admin account username                                                                                                                                                                  |
| `WG_ADMIN_PASSWORD`        | `--admin-password`        | `adminPassword`        | Yes      |                                         | The admin account password                                                                                                                                                                  |
| `WG_PORT`                  | `--port`                  | `port`                 |          | `8000`                                  | The port the web ui will listen on (http)                                                                                                                                                   |
| `WG_EXTERNAL_HOST`         | `--external-host`         | `externalHost`         |          |                                         | The external domain for the server (e.g. https://www.mydomain.com)                                                                                                                          |
| `WG_STORAGE`               | `--storage`               | `storage`              |          | `sqlite3:///data/db.sqlite3`            | A storage backend connection string. See [storage docs](./3-storage.md)                                                                                                                     |
| `WG_DISABLE_METADATA`      | `--disable-metadata`      | `disableMetadata`      |          | `false`                                 | Turn off collection of device metadata logging. Includes last handshake time and RX/TX bytes only.                                                                                          |
| `WG_WIREGUARD_ENABLED`     | `--[no-]wireguard-enabled`     | `wireguard.enabled`    |          | `true`                                  | Enable/disable the wireguard server. Useful for development on non-linux machines.                                                                                                          |
| `WG_WIREGUARD_INTERFACE`   | `--wireguard-interface`   | `wireguard.interface`  |          | `wg0`                                   | The wireguard network interface name                                                                                                                                                        |
| `WG_WIREGUARD_PRIVATE_KEY` | `--wireguard-private-key` | `wireguard.privateKey` | Yes      |                                         | The wireguard private key. This value is required and must be stable. If this value changes all devices must re-register.                                                                   |
| `WG_WIREGUARD_PORT`        | `--wireguard-port`        | `wireguard.port`       |          | `51820`                                 | The wireguard server port (udp)                                                                                                                                                             |
| `WG_VPN_CIDR`              | `--vpn-cidr`              | `vpn.cidr`             |          | `10.44.0.0/24`                          | The VPN network range. VPN clients will be assigned IP addresses in this range.                                                                                                             |
| `WG_VPN_GATEWAY_INTERFACE` | `--vpn-gateway-interface` | `vpn.gatewayInterface` |          | _default gateway interface (e.g. eth0)_ | The VPN gateway interface. VPN client traffic will be forwarded to this interface.                                                                                                          |
| `WG_VPN_ALLOWED_IPS`       | `--vpn-allowed-ips`       | `vpn.allowedIPs`       |          | `0.0.0.0/1, 128.0.0.0/1`                | Allowed IPs that clients may route through this VPN. This will be set in the client's WireGuard connection file and routing is also enforced by the server using iptables.                  |
| `WG_DNS_ENABLED`           | `--[no-]dns-enabled`           | `dns.enabled`          |          | `true`                                  | Enable/disable the embedded DNS proxy server. This is enabled by default and allows VPN clients to avoid DNS leaks by sending all DNS requests to wg-access-server itself.                  |
| `WG_DNS_UPSTREAM`          | `--dns-upstream`          | `dns.upstream`         |          | _resolveconf autodetection or 1.1.1.1_  | The upstream DNS server to proxy DNS requests to. By default the host machine's resolveconf configuration is used to find it's upstream DNS server, otherwise 1.1.1.1 (cloudflare) is used. |

## The Config File (config.yaml)

Here's an example config file to get started with.

```yaml
loglevel: info
storage: sqlite3:///data/db.sqlite3
wireguard:
  privateKey: "<some-key>"
dns:
  upstream:
    - "8.8.8.8"
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v0.3.0]

### Added

- arm64 and arm/v7 docker image support + github actions thanks to [@timtorChen](https://github.com/Place1/wg-access-server/pull/73)

### Changed

- the wireguard private key is now required when the storage backend is persistent (i.e. not `memory://`)
- configuration flags, environment variables and file properties have been refactored for consistency
  * all configuration file properties (excluding auth providers) can now be set via flags and environment variables
  * all environment variables are prefixed with `WG_` to avoid collisions in hosted environments like Kubernetes
  * all flags & environment variables are named consistently
  * **breaking:** no functionality has been removed but you'll need to update any flags/envvars that you're using

### Deprecations

- deprecated support for having no admin account
  * a config error will be thrown in v0.4.0 if an admin account is not configured
  * see the README.md for examples on setting the admin account
- deprecated file:// storage in favour of sqlite3://
  * will be removed in v0.4.0
  * there is now a storage `migrate` command that you can use to move your data to a different storage backend
  * see the docs for migrating your data: https://place1.github.io/wg-access-server/3-storage/#example-file-to-sqlite3

## [0.2.5]

### Added

- Admin users can now delete devices from the "all devices" page (issue [#57](https://github.com/Place1/wg-access-server/issues/57))

### Bug Fixes

- Fixes website routing to solve 404s (issue [#56](https://github.com/Place1/wg-access-server/issues/56))

## [0.2.4]

### Bug Fixes

- Improved config validation and error reporting (issue [#58](https://github.com/Place1/wg-access-server/issues/58) [#61](https://github.com/Place1/wg-access-server/issues/61))

## [0.2.3]

### Added

- Helm chart now supports configuring a LoadBalancer service for the web ui ([@nqngo](https://github.com/Place1/wg-access-server/pull/60))

## [0.2.2]

### Changed

- Changed the default "AllowedIPs" to `0.0.0.0/0`

## [0.2.1]

### Changed

- The "is connected" now shows devices as connected if they've been active within the last 3 minutes
- Improved handling of oidc/gitlab authentication with domain verification when a user hasn't set their email

## [0.2.0]

### Added

- New SQL storage backend supporting SQLite, MySQL and PostgreSQL ([@halkeye](https://github.com/Place1/wg-access-server/pull/37))
- Support for mapping claims from an OIDC auth backend to wg-access-server claims using a simple rule
  syntax ([@halkeye](https://github.com/Place1/wg-access-server/pull/39)). You can use this feature
  to decide which user has the 'admin' claim based on your own OIDC claims.
- The VPN DNS proxy feature can now be disabled using config: `dns.enabled = false`
  - When disabled the `DNS` wireguard config value will be omitted from client wg config files
  - When disabled the DNSasd proxy will not be started server-side (i.e. port 53 won't be used)
- Config options to change the web, wireguard and dns ports.
- Better instructions for connecting a linux device ([@nfg](https://github.com/Place1/wg-access-server/pull/38))
- More helm chart flexibility ([@halkeye](https://github.com/Place1/wg-access-server/pull/33))

### Changes

- The admin UI will now show the device owner's name or email if available.
- The admin UI will now show the auth provider for a given device if more than 1 auth provider is in use.
- Bug fix: upstream dns now correctly configured using resolvconf if not set in config file, flag or envvar.

### Removed

- dns port configuration was removed because wireguard client's only support port 53 for dns

### How to upgrade

- If you've been using the `storage.directory="/some/path"` config value then
  you'll need to update it to `storage=file:///some/path`
- If you've been using the `--storage-directory=/some/path` cli flag then
  you'll need to update it to `--storage="file:///some/path"`
- If you've been using the `STORAGE_DIRECTORY=/some/path` environment variable then
  you'll need to update it to `STORAGE="file:///some/path"`

## [0.1.1]

### Changes

- Helm chart bug fixes and improvements

## [0.1.0]

### Added

- Added support for an admin account. An admin can see all devices registered
  with the server.
- Added support for configuring "AllowedIPs"
- New docker compose example ([@antoniebou13](https://github.com/Place1/wg-access-server/pull/13))
- Added a helm chart
- Added a basic kubernetes quickstart.yaml manifest (based on helm template)
- Added a documentation site based on [mkdocs](https://www.mkdocs.org/). Hosted
  on github pages (still a wip!)

## [0.0.9]

### Changed

- Some UI/UX improvements

## [0.0.8]

### Added

- Added an embedded DNS proxy

### Changed

- Completely re-implemented the auth subsystem to avoid trying to integrate
  with Dex. OIDC, Gitlab and Basic auth are supported.

## [0.0.0] -> [0.0.7]

MVP :)


internal/config/config.go
package config

import (
	"github.com/place1/wg-access-server/pkg/authnz/authconfig"
)

type AppConfig struct {
	// Set the log level.
	// Defaults to "info" (fatal, error, warn, info, debug, trace)
	LogLevel string `yaml:"loglevel"`
	// Set the superadmin username
	// Defaults to "admin"
	AdminUsername string `yaml:"adminUsername"`
	// Set the superadmin password (required)
	AdminPassword string `yaml:"adminPassword"`
	// Port sets the port that the web UI will listen on.
	// Defaults to 8000
	Port int `yaml:"port"`
	// ExternalAddress is the address that clients
	// use to connect to the wireguard interface
	// By default, this will be empty and the web ui
	// will use the current page's origin.
	ExternalHost string `yaml:"externalHost"`
	// The storage backend where device configuration will
	// be persisted.
	// Supports memory:// file:// postgres:// mysql:// sqlite3://
	// Defaults to memory://
	Storage string `yaml:"storage"`
	// DisableMetadata allows you to turn off collection of device
	// metadata including last handshake time & rx/tx bytes
	DisableMetadata bool `yaml:"disableMetadata"`
	// Configure WireGuard related settings
	WireGuard struct {
		// Set this to false to disable the embedded wireguard
		// server. This is useful for development environments
		// on mac and windows where we don't currently support
		// the OS's network stack.
		Enabled bool `yaml:"enabled"`
		// The network interface name of the WireGuard
		// network device.
		// Defaults to wg0
		Interface string `yaml:"interface"`
		// The WireGuard PrivateKey
		// If this value is lost then any existing
		// clients (WireGuard peers) will no longer
		// be able to connect.
		// Clients will either have to manually update
		// their connection configuration or setup
		// their VPN again using the web ui (easier for most people)
		PrivateKey string `yaml:"privateKey"`
		// The WireGuard ListenPort
		// Defaults to 51820
		Port int `yaml:"port"`
	} `yaml:"wireguard"`
	// Configure VPN related settings (networking)
	VPN struct {
		// CIDR configures a network address space
		// that client (WireGuard peers) will be allocated
		// an IP address from
		// defaults to 10.44.0.0/24
		CIDR string `yaml:"cidr"`
		// GatewayInterface will be used in iptable forwarding
		// rules that send VPN traffic from clients to this interface
		// Most use-cases will want this interface to have access
		// to the outside internet
		GatewayInterface string `yaml:"gatewayInterface"`
		// The "AllowedIPs" for VPN clients.
		// This value will be included in client config
		// files and in server-side iptable rules
		// to enforce network access.
		// defaults to ["0.0.0.0/1", "128.0.0.0/1"]
		AllowedIPs []string `yaml:"allowedIPs"`
	}
	// Configure the embeded DNS server
	DNS struct {
		// Enabled allows you to turn on/off
		// the VPN DNS proxy feature.
		// DNS Proxying is enabled by default.
		Enabled bool `yaml:"enabled"`
		// Upstream configures the addresses of upstream
		// DNS servers to which client DNS requests will be sent to.
		// Defaults the host's upstream DNS servers (via resolveconf)
		// or 1.1.1.1 if resolveconf cannot be used.
		// NOTE: currently wg-access-server will only use the first upstream.
		Upstream []string `yaml:"upstream"`
	} `yaml:"dns"`
	// Auth configures optional authentication backends
	// to controll access to the web ui.
	// Devices will be managed on a per-user basis if any
	// auth backends are configured.
	// If no authentication backends are configured then
	// the server will not require any authentication.
	Auth authconfig.AuthConfig `yaml:"auth"`
}
